Upgrade to nom 5

jonhoo opened the pull request:
This patchset builds on #21, and bumps the dependency on `nom` to version 5. The change is a fairly major one, since `nom` moved away from macro-based parsers to combinators based on function callbacks.

Fixes #19 

---

**Old obsolete text:**

The current state of affairs is that this _almost_ compiles. The biggest remaining issue is how to pass the methods of `expr::PRef` in the new function-passing style. The errors should pretty quickly make it obvious what's going on. I'm honestly not sure what the right way to go about this is. The methods that call them may need to be re-written in the sort of "linear" style that `nom 5` favors. [`nom-methods`](https://docs.rs/nom-methods/) may provide some inspiration here. The new `&mut self` style also makes it awkward to have a combinator that may call multiple different methods on `self`.  It may be that we want to return to the old "consume self return Self" strategy to get around this.

I don't know that I'll have the time to finish this PR up any time soon, so @jethrogb if you want to pick this up and run with it, feel free!

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jethrogb/rust-cexpr/22)
<!-- Reviewable:end -->

jonhoo commented:
@Geal I noticed your thumbs-up :p Do you have any good suggestions for how to deal with parser methods that take `&mut self` combined with `alt`?

jonhoo commented:
Apparently (?) the `mut self` wasn't necessary... I just turned them all into `&self` (which allows aliasing across the closures in the branches of `any`, and it just worked.. And all the tests pass. So I guess ready for review!

jonhoo commented:
As a separate observation, since this will be a breaking change, I think we should also take the time to consider any _other_ breaking changes we want to make to make to the API. Maybe @emilio has some ideas given the experience with cexpr in bindgen?

jonhoo commented:
I can also happily report that I tried running the bindgen test suite with this branch of `cexpr` patched in, and all the tests pass with no further modifications!

jrmuizel commented:
Is this ready to be merged?

jonhoo commented:
@jrmuizel It should be -- just waiting on @jethrogb to review.

jethrogb commented on src/literal.rs:
```diff
@@ -166,62 +173,77 @@ fn c_unicode_escape(n: Vec<u8>) -> Option<CChar> {
+    use nom::character::complete::char;
+    use nom::combinator::map;
+    use nom::multi::fold_many0;
+    use nom::sequence::{delimited, preceded};
```
Not a fan of all these uses at the start of functions. Is this canonical nom use now? I'd rather have a bunch of imports at the top.

jethrogb commented on Cargo.toml:
```diff
@@ -13,7 +13,7 @@ keywords = ["C","expression","parser"]
 travis-ci = { repository = "jethrogb/rust-cexpr" }
```
Can you update the version (higher up) to 0.4?

jethrogb commented on src/expr.rs:
```diff
@@ -88,89 +89,80 @@ impl From<Vec<u8>> for EvalResult {
+fn any_token(input: &[Token]) -> CResult<'_, &Token> {
+    if input.is_empty() {
+        let res: CResult<'_, &Token> = Err(crate::nom::Err::Incomplete(Needed::Size(1)));
+        res
```
I think you no longer need to store this in `res` now that this is a function

jethrogb commented on src/expr.rs:
```diff
@@ -297,120 +289,142 @@ fn unary_op(input: (&[u8], EvalResult)) -> Option<EvalResult> {
-			map_opt!(pair!(one_of_punctuation!(["+", "-", "~"]),call_m!(self.unary)),unary_op)
-		)
-	);
+    fn unary<'t>(&'_ self, input: &'t [Token]) -> CResult<'t, EvalResult> {
```
Please write `&self` instead of `&'_ self` (everywhere)

jonhoo replied:
I _think_ it's actually important for it to be `'_` so that the compiler does not assume that the lifetime of `&self` is also included in `'t`. We want the returned value to be able to outlive `&self`. But it could be that my understanding of inferred lifetime bounds is incorrect here?

jethrogb replied:
There are special rules for `&self` if you don't manually specify any lifetime parameters. But an unspecified lifetime parameter is never assumed to be the same as a specified one.

jonhoo commented:
Should be all good now. I also went through and fixed all the clippy lints (which did change some signatures), and I added (minimal) documentation to things that did not have it that you may want to look over. For the future, I also added `#[warn(missing_docs)]`.

jethrogb commented:
bors r+

jonhoo commented:
I'll update https://github.com/rust-lang/rust-bindgen/pull/1735 once `0.4.0` is released to crates.io :)

jethrogb commented:
Done
